cpp/game_main.cpp

```diff
@@ -400,96 +469,27 @@ bool GameMain::on_input(SDL_Event *e) {
 		}
 		break;
 
-	case SDL_KEYUP:
-		switch (((SDL_KeyboardEvent *) e)->keysym.sym) {
-
-		case SDLK_ESCAPE:
-			//stop the game
-			engine.stop();
-			break;
-
-		case SDLK_F1:
-			engine.drawing_huds = !engine.drawing_huds;
-			break;
-
-		case SDLK_F2:
-			engine.get_screenshot_manager().save_screenshot();
-			break;
-
-		case SDLK_F3:
-			engine.drawing_debug_overlay = !engine.drawing_debug_overlay;
-			break;
-
-		case SDLK_F4:
-			this->debug_grid_active = !this->debug_grid_active;
-			break;
-
-		case SDLK_F5:
-			gameio::save(this, "default_save.txt");
-			break;
-
-		case SDLK_F9:
-			gameio::load(this, "default_save.txt");
-			break;
-
-		case SDLK_SPACE:
-			this->terrain->blending_enabled = !terrain->blending_enabled;
-			break;
-
-		case SDLK_F12:
-			if (this->external_profiler.currently_profiling) {
-				this->external_profiler.stop();
-				this->external_profiler.show_results();
-			} else {
-				this->external_profiler.start();
-			}
-
-			break;
-
-		case SDLK_LCTRL:
-			this->ctrl_active = false;
-			break;
-		case SDLK_m:
-			this->construct_mode = !this->construct_mode;
-			break;
-		case SDLK_p:
-			UnitAction::show_debug = !UnitAction::show_debug;
-			break;
-		case SDLK_t:
-			// attempt to train editor selected object
-			if ( this->datamanager.producer_count() > 0 ) {
-				UnitProducer *producer = this->datamanager.get_producer_index(this->editor_current_building);
-				Command cmd(this->players[0], producer);
-				this->selection.all_invoke(cmd);
-			}
-			break;
-		case SDLK_y:
-			this->building_placement = true;
-			break;
-		case SDLK_z:
-			this->use_set_ability = false;
-			break;
-		case SDLK_x:
-			this->use_set_ability = true;
-			this->ability = ability_type::move;
-			break;
-		case SDLK_c:
-			this->use_set_ability = true;
-			this->ability = ability_type::gather;
-			break;
-		}
+	case SDL_KEYUP: {
+		this->keymod = SDL_GetModState();
 
+		SDL_Keycode sym = reinterpret_cast<SDL_KeyboardEvent *>(e)->keysym.sym;
+		keybinds::KeybindManager &keybinds = engine.get_keybind_manager();
+		keybinds.set_key_state(sym, false);
+		keybinds.press(keybinds::key_t(sym, keymod));
 		break;
-	case SDL_KEYDOWN:
-		switch (((SDL_KeyboardEvent *) e)->keysym.sym) {
-		case SDLK_LCTRL:
-			this->ctrl_active = true;
-			break;
-		}
+	}
+
+	case SDL_KEYDOWN: {
+		this->keymod = SDL_GetModState();
 
+		SDL_Keycode sym = reinterpret_cast<SDL_KeyboardEvent *>(e)->keysym.sym;
+		engine.get_keybind_manager().set_key_state(sym, true);
```

Comment:
i'm late for this as the PR is already merged, should the game release all the keys when the window loses focus?